### PR TITLE
Don't set :foreground/background to nil to avoid font warnings in emacs29

### LINKED ISSUE #15862: Warning: (...) :foreground of face (...) nil value is invalid
#### Description :octocat:
I see a lot of warnings like
```
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-insert-face’: nil value is invalid, use ‘unspecified’ instead.
```
(see below for full list)

`git grep :foreground nil` and `git grep :background nil` give me a few instances but none of these seem to be relevant to the warnings. (These instances should probably be fixed nevertheless but that's a different issue.) 

#### Reproduction guide :beetle:
- Start Emacs
- Open a new frame

*Observed behaviour:* :eyes: :broken_heart:
```
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-iedit-insert-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-iedit-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-lisp-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-motion-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-visual-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-evilified-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-replace-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-hybrid-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-emacs-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-insert-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-normal-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-iedit-insert-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-iedit-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-lisp-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-motion-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-visual-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-evilified-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-replace-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-hybrid-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-emacs-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-insert-face’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:foreground’ of face ‘spacemacs-normal-face’: nil value is invalid, use ‘unspecified’ instead.
```

*Expected behaviour:* :heart: :smile:
No warnings

#### System Info :computer:
- OS: gnu/linux
- Emacs: 29.0.60
- Spacemacs: 0.999.0
- Spacemacs branch: nil (rev. 025e9f7e3)
- Graphic display: t
- Running in daemon: t
- Distribution: spacemacs
- Editing style: vim
- Completion: ivy
- Layers:
```elisp
(better-defaults
 (unicode-fonts :variables unicode-fonts-enable-ligatures t unicode-fonts-ligature-modes
                '(prog-mode text-mode)
                unicode-fonts-ligature-set
                '("<---" "<--" "<<-" "<-" "->" "-->" "--->" "<->" "<-->" "<--->" "<---->" "<!--" "<==" "<===" "<=" "=>" "=>>" "==>" "===>" ">=" "<=>" "<==>" "<===>" "<====>" "<!---" "<~~" "<~" "~>" "~~>" "::" ":::" "==" "!=" "===" "!==" ":=" ":-" ":+" "<*" "<*>" "*>" "<|" "<|>" "|>" "+:" "-:" "=:" "<******>" "++" "+++"))
 theming
 (ivy :variables ivy-enable-advanced-buffer-information t)
 (auto-completion :disabled-for org)
 spell-checking syntax-checking treemacs shell
 (ibuffer :variables ibuffer-group-buffers-by 'projects)
 (org :variables org-enable-github-support t org-enable-roam-support t org-enable-roam-ui t org-enable-sticky-header t org-want-todo-bindings t)
 deft
 (latex :variables latex-enable-auto-fill nil lsp-headerline-breadcrumb-enable nil lsp-enable-imenu nil lsp-ui-doc-enable nil lsp-eldoc-enable-hover nil)
 bibtex
 (languagetool :variables langtool-java-classpath "/usr/share/languagetool:/usr/share/java/languagetool/*")
 (lsp :variables lsp-rust-server 'rust-analyzer lsp-enable-indentation nil lsp-enable-on-type-formatting nil)
 (c-c++ :variables c-c++-backend 'lsp-ccls c-c++-lsp-enable-semantic-highlight t lsp-lens-enable nil)
 ansible csv html
 (shell-scripts :variables shell-scripts-backend nil)
 javascript systemd emacs-lisp git
 (version-control :variables version-control-diff-side 'left)
 markdown major-modes yaml lua rust php python)
```
- System configuration features: ACL CAIRO DBUS FREETYPE GIF GLIB GMP GNUTLS GSETTINGS HARFBUZZ JPEG JSON LCMS2 LIBSYSTEMD LIBXML2 MODULES NATIVE_COMP NOTIFY INOTIFY PDUMPER PGTK PNG RSVG SECCOMP SQLITE3 THREADS TIFF TOOLKIT_SCROLL_BARS TREE_SITTER WEBP XIM GTK3 ZLIB

### PR DESCRIPTION
**Pull request by real-or-random**
`:foreground nil` (or `:background nil`) are illegal values and generate warnings in emacs 29.

We can't set to `unspecified` either because user code is not supposed to do this:
 - https://github.com/emacs-mirror/emacs/commit/824a139434855919073d6ea3c93d57ebd24fc75c
 - https://debbugs.gnu.org/cgi/bugreport.cgi?bug=51595#15

This eliminates all warnings that I see with my config. There may be more instances of setting to `nil` in the code base. We'll find them once they generate warnings for others.

Fixes #15862.

cc @lebensterben

### CODE REVIEW

**Comment by lebensterben** [comment]
unfortunately it's invalid by changing

```elisp
(set-face-attribute face frame :foreground nil)
```

to
```elisp
(set-face-attribute face frame)
```

this results in a no-op. (I've checked the source code https://github.com/emacs-mirror/emacs/blob/2f05f48918ecf6e59b330709b09a36f81f7b6c48/src/xfaces.c#L3064)

We still need to replace `nil` by `'unspecified`, as also used in
https://github.com/emacs-mirror/emacs/blob/2f05f48918ecf6e59b330709b09a36f81f7b6c48/lisp/faces.el#L969-L976

**Comment by lebensterben** [comment]
Try the following two scenarios:

First,

```elisp
(defface foo '((t :foreground unspecified)) "DOC...")
(customize-face foo)
```

This hits the bug.

Try setting the foreground to a valid color and it **seems** to work.

```elisp
(set-face-attribute 'foo nil :foreground "LightBlue")
(customize-face foo)
```

But if you ever setting the foreground to `'unspecified` again, it will break Customization again

```elisp
(set-face-attribute 'foo nil :foreground 'unspecified)
(customize-face foo)
```

---

Second,

```elisp
(defface bar '((t)) "DOC...")
(customize-face bar)
```

it shows a face with empty attributes.

Try setting any valid color or `'unspecified` as its foreground color and it never breaks Customization.

---

Summary:

It's not erroneous to use `'unspecified` as attribute value in `set-face-attribute` form. But it's wrong in `defface`!

**Comment by lebensterben** [comment]
A similar PR: https://github.com/doomemacs/themes/commit/aff98a6787a17a0ffad72f8d99c17e4a2e724754

They replaced all `nil` by `'unspecified` and I didn't notice anything wrong.

**Comment by real-or-random** [comment]
Updated.

edit: Oh damn, I also need to fix the second commit... I'll try to do that tomorrow

File: core/libs/spacemacs-theme/spacemacs-common.el [hunk 1]
```diff
@@ -277,11 +277,11 @@ to 'auto, tags may not be properly aligned. "
      `(centaur-tabs-modified-marker-unselected ((,class (:inherit 'centaur-tabs-unselected :foreground,keyword))))
 
 ;;;;; cider
-     `(cider-enlightened ((,class (:background nil :box (:color ,yellow :line-width -1 :style nil) :foreground ,yellow))))
+     `(cider-enlightened ((,class (:box (:color ,yellow :line-width -1 :style nil) :foreground ,yellow))))
```
Comment (lebensterben) [inline core/libs/spacemacs-theme/spacemacs-common.el:280]:
```suggestion
     `(cider-enlightened ((,class (:background unspecified :box (:color ,yellow :line-width -1 :style nil) :foreground ,yellow))))
```

**Comment by lebensterben** [comment]
core/libs/spacemacs-themes should not be touched here.

you can make PR to https://github.com/nashamri/spacemacs-theme

File: layers/+themes/theming/README.org [hunk 2]
```diff
@@ -48,8 +48,7 @@ It should be a list of the following form:
           '((monokai
              ;; Font locking
              (font-lock-comment-face :slant italic)
-             (web-mode-html-attr-name-face :inherit font-lock-variable-name-face
-                                           :foreground nil)
+             (web-mode-html-attr-name-face :inherit font-lock-variable-name-face)
```
Comment (lebensterben) [inline layers/+themes/theming/README.org]:
```suggestion
             (web-mode-html-attr-name-face :inherit font-lock-variable-name-face
                                           :foreground unspecified)
```

File: layers/+spacemacs/spacemacs-editing/funcs.el [hunk 3]
```diff
@@ -114,9 +114,7 @@ If `global' is non-nil activate the respective global mode."
 
 (defun spacemacs//adaptive-smartparent-pair-overlay-face ()
   (set-face-attribute 'sp-pair-overlay-face nil
-                      :inherit 'lazy-highlight
-                      :background nil
-                      :foreground nil))
+                      :inherit 'lazy-highlight))
```
Comment (lebensterben) [inline layers/+spacemacs/spacemacs-editing/funcs.el]:
```suggestion
                      :inherit 'lazy-highlight
                      :background 'unspecified
                      :foreground 'unspecified))
```

File: layers/+spacemacs/spacemacs-defaults/packages.el [hunk 4]
```diff
@@ -527,13 +527,10 @@
     :config
     (progn
       (set-face-attribute 'whitespace-space nil
```
Comment (lebensterben) [inline layers/+spacemacs/spacemacs-defaults/packages.el:529]:
```suggestion
      (set-face-attribute 'whitespace-space nil
                          :background 'unspecified
```

File: layers/+spacemacs/spacemacs-defaults/packages.el [hunk 5]
```diff
@@ -527,13 +527,10 @@
...
-                          :background nil
                           :foreground (face-attribute 'font-lock-warning-face
                                                       :foreground))
-      (set-face-attribute 'whitespace-tab nil
-                          :background nil)
-      (set-face-attribute 'whitespace-indentation nil
-                          :background nil)
+      (set-face-attribute 'whitespace-tab nil)
```
Comment (lebensterben) [inline layers/+spacemacs/spacemacs-defaults/packages.el]:
```suggestion
      (set-face-attribute 'whitespace-tab nil
                          :background 'unspecified)
```

File: layers/+spacemacs/spacemacs-defaults/packages.el [hunk 6]
```diff
@@ -527,13 +527,10 @@
...
+      (set-face-attribute 'whitespace-indentation nil)
```
Comment (lebensterben) [inline layers/+spacemacs/spacemacs-defaults/packages.el]:
```suggestion
      (set-face-attribute 'whitespace-indentation nil
                          :background 'unspecified)
```

File: layers/+spacemacs/spacemacs-completion/funcs.el [hunk 7]
```diff
@@ -43,8 +43,6 @@
   "Set the face of diretories for `.' and `..'"
   (set-face-attribute 'helm-ff-dotted-directory
                       nil
```
Comment (lebensterben) [inline layers/+spacemacs/spacemacs-completion/funcs.el:45]:
```suggestion
                      nil
                      :foreground 'unspecified
                      :background 'unspecified
```

**Review by lebensterben — CHANGES_REQUESTED** [review]

**Comment by eeshugerman** [comment]
Thanks for working on this @real-or-random!

Please also change stuff like `:background ,(when foo bar)` to `:background ,(if foo bar 'unspecified)`.

Here are some warnings I'm seeing that this would address:
```
Warning: setting attribute ‘:background’ of face ‘org-level-4’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:background’ of face ‘org-level-3’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:background’ of face ‘org-level-2’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:background’ of face ‘org-level-1’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:background’ of face ‘markdown-header-face-4’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:background’ of face ‘markdown-header-face-3’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:background’ of face ‘markdown-header-face-2’: nil value is invalid, use ‘unspecified’ instead.
Warning: setting attribute ‘:background’ of face ‘markdown-header-face-1’: nil value is invalid, use ‘unspecified’ instead.
```

**Comment by real-or-random** [comment]
> core/libs/spacemacs-themes should not be touched here.
> 
> you can make PR to [nashamri/spacemacs-theme](https://github.com/nashamri/spacemacs-theme)

Ok! Ready for review. I figured out that `(face-attribute face :background)` is often more robust for these use cases because it will simply return `unspecified` instead of mapping it to `nil`.

> Thanks for working on this @real-or-random!
>
> Please also change stuff like :background ,(when foo bar) to :background ,(if foo bar 'unspecified).

@eeshugerman 
This also needs to be done in spacemacs-theme. Are you willing to give it a try? I'm not using spacemacs-theme, so testing is probably easier for you.

File: layers/+distributions/spacemacs-bootstrap/funcs.el [hunk 8]
```diff
@@ -67,19 +67,20 @@ For evil states that also need an entry to `spacemacs-evil-cursors' use
...
 
 (defun spacemacs/set-state-faces ()
   (cl-loop for (state color cursor) in spacemacs-evil-cursors
            do
-           (set-face-attribute (intern (format "spacemacs-%s-face" state))
-                               nil
-                               :foreground (face-background 'mode-line))))
+           (set-face-attribute (spacemacs/state-color-face (intern state)) nil
+                               :foreground
+                               (face-attribute 'mode-line :background))))
```
Comment (lebensterben) [inline layers/+distributions/spacemacs-bootstrap/funcs.el]: you can acquire this value outside of this loop so it only needs to be taken for once.

**Comment by lebensterben** [comment]
I also happen to find another issue:
`defface` now requires at least three arguments, face name, face specification, and documentation. But we only supplied it with 2!

---

Also, I don't see why we have

```elisp
(eval `(defface ..))
```

instead of

```elisp
(defface ..)
```

directly. Can you also try that?

File: layers/+distributions/spacemacs-bootstrap/funcs.el (hunk 8 above)
Comment (real-or-random) [reply layers/+distributions/spacemacs-bootstrap/funcs.el]: done

**Comment by real-or-random** [comment]
Arguments of `defface`:

Where? I don't think I'm touching code that gets this wrong.

----

`eval` and `defface`:
Though this is beyond my lisp skills, using `eval` seems to be a common thing because `defface` is a macro (see https://stackoverflow.com/questions/20000531/how-to-define-multiple-emacs-faces). We use `eval` with `defface` a few other times in the code base. It's not the goal of this PR to change this.

**Comment by lebensterben** [comment]
@real-or-random 

I just made a change that I replaced

```
(eval `(defface ..))
```

by

```
`(defface ..)
```

(by changing the function `spacemacs/define-evil-state-face` to a macro `spacemacs|define-evil-state-face`)

Can you pull in the changes and verify whether that works?

**Comment by lebensterben** [comment]
Thx

**Pull request merged**